### PR TITLE
Implement Syntax Highlight inside macro call

### PR DESCRIPTION
**Pull request by edwin0cheng**

### CODE REVIEW

File: crates/ra_ide/src/syntax_highlighting.rs [hunk 1]
```diff
@@ -72,121 +76,186 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
...
-            use std::{collections::hash_map::DefaultHasher, hash::Hasher};
+    let mut sb = SourceBinder::new(db);
+    let mut bindings_shadow_count: FxHashMap<Name, u32> = FxHashMap::default();
+    let mut res = Vec::new();
+    let analyzer = sb.analyze(InFile::new(file_id.into(), &root), None);
 
-            let mut hasher = DefaultHasher::new();
-            x.hash(&mut hasher);
-            hasher.finish()
+    let mut in_macro_call = None;
```
Comment (matklad) [inline crates/ra_ide/src/syntax_highlighting.rs]: Will this handle recursive macro invocations correctly? I haven't looked closely, but it seems like the currect code can remeber only *one* macro?

Comment (edwin0cheng) [reply crates/ra_ide/src/syntax_highlighting.rs]: Here we only handle the `AstNode` from root source such that the recursive macro invocations will be treated as TOKENTREE node inside MacroCall Node, for examples:

```
bar!(foo!())
```

```
MACRO_CALL@[14763; 14776)
    PATH@[14763; 14766)
      PATH_SEGMENT@[14763; 14766)
        NAME_REF@[14763; 14766)
          IDENT@[14763; 14766) "bar"
    EXCL@[14766; 14767) "!"
    TOKEN_TREE@[14767; 14775)
      L_PAREN@[14767; 14768) "("
      IDENT@[14768; 14771) "foo"
      EXCL@[14771; 14772) "!"
      TOKEN_TREE@[14772; 14774)
        L_PAREN@[14772; 14773) "("
        R_PAREN@[14773; 14774) ")"
      R_PAREN@[14774; 14775) ")"
    SEMI@[14775; 14776) ";"
```

Comment (edwin0cheng) [reply crates/ra_ide/src/syntax_highlighting.rs]: So we don't need to handle recursive macro call here, and will handle it inside `highlight_token_tree` :

 https://github.com/rust-analyzer/rust-analyzer/blob/f320af4d63302d2933b37794826f705f13caf8a0/crates/ra_ide/src/syntax_highlighting.rs#L163

**Comment by matklad** [comment]
Yeah, this all looks super reasonable to me, 

bors r+

**Pull request merged**